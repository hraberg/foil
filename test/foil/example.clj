(ns foil.example
  (:require [cstdio]
            [cmath]
            [algorithm]
            [functional]
            [iostream]))

(defrecord Point [^int x ^int y])

(def foo 3.14)
(def ^:dynamic *pi* 3.14)

($code "#define _PLUS_ std::plus<>{}")
($code "#define _MINUS_ std::minus<>{}")
($code "#define _STAR_ std::multiplies<>{}")
($code "#define _SLASH_ std::divides<>{}")
($code "#define _PERCENT_ std::modulus<>{}")
($code "#define _EQ_ std::equal_to<>{}")
($code "#define _BANG__EQ_ std::not_equal_to<>{}")
($code "#define _GT_ std::greater<>{}")
($code "#define _LT_ std::less<>{}")
($code "#define _GT__EQ_ std::greater_equal<>{}")
($code "#define _LT_EQ_ std::less_equal<>{}")
($code "#define _AMPERSAND__AMPERSAND_ std::logical_and<>{}")
($code "#define _BAR__BAR_ std::logical_or<>{}")
($code "#define _BANG_ std::logical_not<>{}")
($code "#define _AMPERSAND__ std::bit_and<>{}")
($code "#define _BAR_ std::bit_or<>{}")
($code "#define _CARET_ std::bit_xor<>{}")
($code "#define _TILDE_ std::not_fn<>{}")

($code "
    template <typename Fn, typename T, typename... TRest, template <typename...>typename Coll>
    auto map(const Fn& f, const Coll<T, TRest...>& coll) {
        Coll<decltype(f(std::declval<T>()))> acc;
        for (const auto& x : coll) {
            acc.push_back(f(x));
        }
        return acc;
    }")

($code "
    template <typename Pred, typename Coll>
    auto filter(const Pred& pred, const Coll& coll) {
        Coll acc;
        for (const auto& x : coll) {
            if (pred(x)) {
                acc.push_back(x);
            }
        }
        return acc;
    }")

($code "
    template <typename Fn, typename Coll, typename Val>
    auto reduce(const Fn& f, const Val& val, const Coll& coll) {
        Val acc = val;
        for (const auto& x : coll) {
            acc = f(acc, x);
        }
        return acc;
    }")

($code "
    template <typename Fn, typename Coll>
    auto reduce(const Fn& f, const Coll& coll) {
        typename Coll::value_type acc = coll.front();
        auto first = true;
        for (const auto& x : coll) {
            if (first) {
                first = false;
                continue;
            }
            acc = f(acc, x);
        }
        return acc;
    }")

(defn println ^void [^std::string x]
  (<< (<< (<< std::cout x) " str") std::endl))

(defn println ^void [x]
  (<< (<< std::cout x) std::endl))

(defn -main []
  (let [^:mut x 0
        pt (Point. -1 2)
        t 3
        ^:mut a ^int [2 t]
        m ^std::string {:foo "bar"}
        s ^int #{1 2}
        l ^int '(7 8)
        my-fun (fn [x] (printf (.c_str "hello, world %d\n") x))]

    (aset a 0 4)
    (printf (.c_str "%d %d %s %.2f %lu %lu\n") (.-x pt) (aget a 0) (.c_str (.at m :foo)) foo (.count s 1) (.count s 3))

    (while (< x 10)
      (set! x (+ x 1)))

    (doseq [x s
            y a]
      (printf (.c_str "%d %d\n") x y))

    (dotimes [n 3]
      (println n))

    (my-fun x)

    (println *pi*)
    (binding [*pi* -1]
      (println *pi*))
    (println *pi*)

    (println (if (= 10 x)
               "ten"
               "not ten"))

    (println (cond
               (= 10 x)
               "ten"

               (= 20 x)
               "not ten"))

    (println 42)

    (let [s "foo"]
      (println s))

    (println (.front l))

    (println (sin *pi*))

    (println (loop [n 0]
               (if (< n 3)
                 (recur (inc n))
                 n)))

    (let [^:mut x ^int []
          ^:mut z ^int []]

      (std::transform (.begin a)
                      (.end a)
                      (std::back_inserter x)
                      #(inc %))


      (std::for_each (.begin x)
                     (.end x)
                     #(println %))

      (println (std::accumulate (.begin x)
                                (.end x)
                                0
                                +))

      (let [^:mut zz ^int []]
        (std::copy_if (.begin x)
                      (.end x)
                      (std::back_inserter zz)
                      #(= (mod % 2) 0))
        (doseq [x zz]
          (println x))))

    (let [xs (map #(inc %) a)]
      (doseq [x xs]
        (println x))

      (println (reduce + xs))

      (doseq [x (filter #(= (mod % 2) 0)
                        (map #(inc %) a))]
        (println x)))
    0)
  )
