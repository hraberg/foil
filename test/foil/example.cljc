(ns example
  (:require [cstdio]
            [cmath]
            [algorithm]
            [functional]
            [iostream]
            [memory]))

(defrecord Point ^{:tpl "<typename X = int>"} [^X x ^int y])

(def foo 3.14)
(def ^:dynamic *pi* 3.14)

(defn test-println
  (^void [^std::string x]
   (<< std::cout x " str" std::endl))
  (^void [x]
   (println x)))

(defn -main []
  (let [^:mut x 0
        pt ^int (Point. -1 2)
        t 3
        ^:mut a ^int [2 t]
        m ^std::string {:foo "bar"}
        s ^int #{1 2}
        l ^int (list 7 8)
        my-fun (fn [x] (printf (.c_str "hello, world %d\n") x))
        ^:mut ss ^int [2 3 1]]

    (println ((partial + 2) 4))
    (println)

    (doseq [x (sort ss)]
      (println x))

    (let [sp1 ^std::vector<int> (std::make_shared ^int [1 2])
          sp2 (identity sp1)]
      (println (aget @sp1 0))
      (aset (* sp1) 0 0)
      (println (= @sp2 @sp1))
      (aset (* sp2) 1 3)
      (println (= @sp2 @sp1))
      (println (.use_count sp1)))

    (println (identity 5))

    (println (min 3 2))

    (println (first (next l)))
    (println (first l))
    (println (second l))
    (println (count l))
    (println (first (cons 6 l)))
    (println (= l ^int '(7 8)))
    (println (= l (cons 6 l)))
    (println (count (next l)))
    (println (count (next (next l))))
    (println (count (next (next (next l)))))

    (println @(first-opt l))
    (println (boolean (first-opt ^int ())))

    (println (nth ^std::string ["hello" "world"] 1 "?"))
    (println (nth ^std::string ["hello" "world"] 3 "?"))

    (aset a 0 4)
    (printf (.c_str "%d %d %s %.2f %lu %lu\n") (.-x pt) (aget a 0) (.c_str (get m :foo)) foo (contains? s 1) (contains? s 3))

    (doseq [x m]
      (println (key x))
      (println (val x)))

    (doseq [x (keys m)]
      (println x))

    (while (< x 10)
      (set! x (+ x 1)))

    (doseq [x s
            y a]
      (printf (.c_str "%d %d\n") x y))

    (dotimes [n 3]
      (test-println n))

    (my-fun x)

    (println (count (concat l s)))
    (println (= ^int '(8 7) (reverse l)))

    (println (every? pos? l))
    (println (every? neg? l))
    (println (= ^int [8] (drop 1 l)))
    (println (= ^int [7] (take 1 l)))

    (let [^:mut xs ^int [7]]
      (println (count (empty! xs)))
      (println (count xs)))

    (let [xs ^int [7]]
      (println (count (empty xs)))
      (println (count xs)))

    (doseq [x (map-indexed (fn [idx _]
                             idx) l)]
      (println x))

    (println (contains? (set l) 7))
    (doseq [x ^int (sorted-set 8 7)]
      (println x))

    (doseq [x ^int (sorted-set-by > 8 7)]
      (println x))

    (doseq [x ^"std::string,std::string" (hash-map ^std::string ["foo" "bar"])]
      (println (key x))
      (println (val x)))

    (doseq [x ^"int,int" (sorted-map-by >
                                        ^int [1 2]
                                        ^int [3 4])]
      (println (key x))
      (println (val x)))

    (test-println *pi*)
    (binding [*pi* -1]
      (test-println *pi*))
    (test-println *pi*)

    (test-println (if (= 10 x)
                    "ten"
                    "not ten"))

    (test-println (cond
                    (= 10 x)
                    "ten"

                    (= 20 x)
                    "not ten"))

    (test-println 42)

    (let [s "foo"]
      (test-println s))

    (test-println (first l))

    (test-println (sin *pi*))

    (test-println (loop [n 0]
                    (if (< n 3)
                      (recur (inc n))
                      n)))

    (let [^:mut x ^int []
          ^:mut z ^int []]

      (std::transform (.begin a)
                      (.end a)
                      (std::back_inserter x)
                      inc)

      (std::for_each (.begin x)
                     (.end x)
                     test-println)

      (test-println (std::accumulate (.begin x)
                                     (.end x)
                                     0
                                     +))

      (let [^:mut zz ^int []]
        (std::copy_if (.begin x)
                      (.end x)
                      (std::back_inserter zz)
                      #(= (mod % 2) 0))
        (doseq [x zz]
          (test-println x))))

    (let [xs (map inc a)]
      (doseq [x xs]
        (test-println x))

      (test-println (reduce + xs))

      (doseq [x (->> a
                     (map inc)
                     (filter #(= (mod % 2) 0)))]
        (test-println x)))

    (let [^:mut ^std::atomic<int> at 2]
      (println @at)
      (swap! at + 3)
      (println @at))

    0))
