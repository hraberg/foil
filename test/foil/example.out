6

1
2
3
1
true
true
2
5
2
8
7
8
2
6
true
false
1
0
0
7
false
world
?
-1 4 bar 3.14 1 0
:foo
bar
:foo
2 4
2 3
1 4
1 3
0
1
2
hello, world 10
4
true
true
false
true
true
0
0
0
1
0
1
true
7
8
8
7
foo
bar
3
4
1
2
3.14
-1
3.14
ten str
ten str
42
foo str
7
0.00159265
3
5
4
9
4
5
4
9
4
2
5
